modyn/supervisor/internal/pipeline_executor/pipeline_executor.py

```diff
@@ -1,554 +1,855 @@
-        assert self.pipeline_id is not None, "_run_training called without a registered pipeline."
-        self._update_pipeline_stage_and_enqueue_msg(
-            PipelineStage.RUN_TRAINING, MsgType.ID, id_submsg(IdType.TRIGGER, trigger_id)
-        )
-        logger.info(f"Running training for trigger {trigger_id}")
+        Args:
+            s: Execution state of the pipeline executor.
+            log: Log of the current stage.
+            batch: List of tuples (key, timestamp, label) of new data points.
 
-        self._sw.start("train", overwrite=True)
-        num_samples_to_pass_per_trigger = self.pipeline_config["training"].get("num_samples_to_pass", [])
-        current_trigger_index = len(self.triggers)
-        if current_trigger_index <= len(num_samples_to_pass_per_trigger) - 1:
-            num_samples_to_pass = num_samples_to_pass_per_trigger[current_trigger_index]
-        else:
-            num_samples_to_pass = None
+        Returns:
+            List of indexes of data points that caused a trigger.
+        """
 
-        self.current_training_id = self.grpc.start_training(
-            self.pipeline_id, trigger_id, self.pipeline_config, self.previous_model_id, num_samples_to_pass
-        )
+        # Evaluate trigger policy and inform selector
+        lazy_trigger_indexes = self._evaluate_trigger_policies(s, self.logs, batch)
 
-        self.stage = PipelineStage.WAIT_FOR_TRAINING_COMPLETION
-        trainer_log = self.grpc.wait_for_training_completion(self.current_training_id, self.pipeline_id, trigger_id)
+        # Execute triggers within batch (training & evaluation subpipelines)
+        executed_triggers = self._execute_triggers(s, self.logs, batch, lazy_trigger_indexes)
 
-        if trigger_id not in self.pipeline_log["supervisor"]["triggers"]:
-            self.pipeline_log["supervisor"]["triggers"][trigger_id] = {}  # can happen in tests
+        # Inform selector about remaining data
+        self._inform_selector_remaining_data(s, self.logs, batch, executed_triggers)
 
-        self.pipeline_log["supervisor"]["triggers"][trigger_id]["total_trainer_time"] = self._sw.stop()
-        self.pipeline_log["supervisor"]["triggers"][trigger_id]["trainer_log"] = trainer_log
+        return executed_triggers
 
-        self._update_pipeline_stage_and_enqueue_msg(
-            PipelineStage.STORE_TRAINED_MODEL, MsgType.ID, id_submsg(IdType.TRIGGER, trigger_id)
-        )
-        # We store the trained model for evaluation in any case.
-        self._sw.start("store_trained_model", overwrite=True)
-        model_id = self.grpc.store_trained_model(self.current_training_id)
-        self.trigger.inform_previous_model(model_id)
-        self.pipeline_log["supervisor"]["triggers"][trigger_id]["store_trained_model_time"] = self._sw.stop()
+    @pipeline_stage(PipelineType.NEW_BATCH, PipelineStage.EVALUATE_TRIGGER_POLICIES, track=True)
+    def _evaluate_trigger_policies(
```

Comment:
nit: is there any reason this is plural? conceptually, there is just a single policy

Reply:
It would be possible to evaluate multiple policies before deciding. However, the final decision could then be again seen as a single, higher-order (aggregation) policy. So singular makes sense I guess.